// @clerk/react-router/dist/server/clerkMiddleware.js
import { patchRequest } from "./utils.js";
import { loadOptions } from "./loadOptions.js";
import { clerkClient } from "./clerkClient.js";
import { resolveKeysWithKeylessFallback } from "./keyless/utils.js";
import { createContext } from "react-router";
import { AuthStatus, constants, createClerkRequest } from "@clerk/backend/internal";
import { handleNetlifyCacheInDevInstance } from "@clerk/shared/netlifyCacheHandler";

//#region src/server/clerkMiddleware.ts
const authFnContext = createContext(null);
const requestStateContext = createContext(null);
/**
* Middleware that integrates Clerk authentication into your React Router application.
* It checks the request's cookies and headers for a session JWT and, if found,
* attaches the Auth object to a context.
*
* @example
* // react-router.config.ts
* export default {
*   future: {
*     v8_middleware: true,
*   },
* }
*
* // root.tsx
* export const middleware: Route.MiddlewareFunction[] = [clerkMiddleware()]
*/
const clerkMiddleware = (options) => {
	return async (args, next) => {
		const clerkRequest = createClerkRequest(patchRequest(args.request));
		const loadedOptions = loadOptions(args, options);
		const { publishableKey, secretKey, claimUrl: __keylessClaimUrl, apiKeysUrl: __keylessApiKeysUrl } = await resolveKeysWithKeylessFallback(loadedOptions.publishableKey, loadedOptions.secretKey, args, options);
		if (publishableKey) loadedOptions.publishableKey = publishableKey;
		if (secretKey) loadedOptions.secretKey = secretKey;
		const { apiUrl, jwtKey, proxyUrl, isSatellite, domain, machineSecretKey, audience, authorizedParties, signInUrl, signUpUrl, organizationSyncOptions } = loadedOptions;
		const requestState = await clerkClient(args, options).authenticateRequest(clerkRequest, {
			apiUrl,
			secretKey: loadedOptions.secretKey,
			jwtKey,
			proxyUrl,
			isSatellite,
			domain,
			publishableKey: loadedOptions.publishableKey,
			machineSecretKey,
			audience,
			authorizedParties,
			organizationSyncOptions,
			signInUrl,
			signUpUrl,
			acceptsToken: "any"
		});
		const locationHeader = requestState.headers.get(constants.Headers.Location);
		if (locationHeader) {
			handleNetlifyCacheInDevInstance({
				locationHeader,
				requestStateHeaders: requestState.headers,
				publishableKey: requestState.publishableKey
			});
			return new Response(null, {
				status: 307,
				headers: requestState.headers
			});
		}
		if (requestState.status === AuthStatus.Handshake) throw new Error("Clerk: handshake status without redirect");
		args.context.set(authFnContext, (opts) => requestState.toAuth(opts));
		args.context.set(requestStateContext, {
			requestState,
			additionalState: {
				__keylessClaimUrl,
				__keylessApiKeysUrl,
				signInForceRedirectUrl: loadedOptions.signInForceRedirectUrl,
				signUpForceRedirectUrl: loadedOptions.signUpForceRedirectUrl,
				signInFallbackRedirectUrl: loadedOptions.signInFallbackRedirectUrl,
				signUpFallbackRedirectUrl: loadedOptions.signUpFallbackRedirectUrl
			}
		});
		const response = await next();
		if (requestState.headers) requestState.headers.forEach((value, key) => {
			response.headers.append(key, value);
		});
		return response;
	};
};

//#endregion
export { authFnContext, clerkMiddleware, requestStateContext };
