// @clerk/react-router/dist/server/utils.js
import { getPublicEnvVariables } from "../utils/env.js";
import { canUseKeyless } from "../utils/feature-flags.js";
import { constants, debugRequestState } from "@clerk/backend/internal";
import { parse } from "cookie";

//#region src/server/utils.ts
function isResponse(value) {
	return value != null && typeof value.status === "number" && typeof value.statusText === "string" && typeof value.headers === "object" && typeof value.body !== "undefined";
}
function isDataWithResponseInit(value) {
	return typeof value === "object" && value != null && "type" in value && "data" in value && "init" in value && value.type === "DataWithResponseInit";
}
function isRedirect(res) {
	return res.status >= 300 && res.status < 400;
}
const parseCookies = (req) => {
	return parse(req.headers.get("cookie") || "");
};
function assertValidHandlerResult(val, error) {
	if (val !== null && typeof val !== "object" || Array.isArray(val)) throw new Error(error || "");
}
/**
* `get` and `set` properties will only be available if v8_middleware flag is enabled
* See: https://reactrouter.com/upgrading/future#futurev8_middleware
*/
const IsOptIntoMiddleware = (context) => {
	return "get" in context && "set" in context;
};
const injectRequestStateIntoResponse = async (response, requestState, context, additionalStateOptions = {}, includeClerkHeaders = false) => {
	const clone = new Response(response.body, response);
	const data = await clone.json();
	const { clerkState, headers } = getResponseClerkState(requestState, context, additionalStateOptions);
	clone.headers.set(constants.Headers.ContentType, constants.ContentTypes.Json);
	if (includeClerkHeaders) headers.forEach((value, key) => {
		clone.headers.append(key, value);
	});
	return Response.json({
		...data || {},
		...clerkState
	}, clone);
};
/**
* Returns the clerk state object and observability headers to be injected into a loader response.
*
* @internal
*/
function getResponseClerkState(requestState, context, additionalStateOptions = {}) {
	const { reason, message, isSignedIn, ...rest } = requestState;
	const envVars = getPublicEnvVariables(context);
	const { __keylessClaimUrl, __keylessApiKeysUrl, ...redirectUrlOptions } = additionalStateOptions;
	const baseState = {
		__clerk_ssr_state: rest.toAuth(),
		__publishableKey: requestState.publishableKey,
		__proxyUrl: requestState.proxyUrl,
		__domain: requestState.domain,
		__isSatellite: requestState.isSatellite,
		__signInUrl: requestState.signInUrl,
		__signUpUrl: requestState.signUpUrl,
		__signInForceRedirectUrl: redirectUrlOptions.signInForceRedirectUrl,
		__signUpForceRedirectUrl: redirectUrlOptions.signUpForceRedirectUrl,
		__signInFallbackRedirectUrl: redirectUrlOptions.signInFallbackRedirectUrl,
		__signUpFallbackRedirectUrl: redirectUrlOptions.signUpFallbackRedirectUrl,
		__clerk_debug: debugRequestState(requestState),
		__clerkJSUrl: envVars.clerkJsUrl,
		__clerkJSVersion: envVars.clerkJsVersion,
		__clerkUIUrl: envVars.clerkUIUrl,
		__clerkUIVersion: envVars.clerkUIVersion,
		__prefetchUI: envVars.prefetchUI,
		__telemetryDisabled: envVars.telemetryDisabled,
		__telemetryDebug: envVars.telemetryDebug,
		__unsafeDisableDevelopmentModeConsoleWarning: envVars.unsafeDisableDevelopmentModeConsoleWarning
	};
	if (canUseKeyless && __keylessClaimUrl) {
		baseState.__keylessClaimUrl = __keylessClaimUrl;
		baseState.__keylessApiKeysUrl = __keylessApiKeysUrl;
	}
	return {
		clerkState: wrapWithClerkState(baseState),
		headers: requestState.headers
	};
}
/**
* Wraps obscured clerk internals with a readable `clerkState` key.
* This is intended to be passed by the user into <ClerkProvider>
*
* @internal
*/
const wrapWithClerkState = (data) => {
	return { clerkState: { __internal_clerk_state: { ...data } } };
};
/**
* Patches request to avoid duplex issues with unidici
* For more information, see:
* https://github.com/nodejs/node/issues/46221
* https://github.com/whatwg/fetch/pull/1457
* @internal
*/
const patchRequest = (request) => {
	const clonedRequest = new Request(request.url, {
		headers: request.headers,
		method: request.method,
		redirect: request.redirect,
		cache: request.cache
	});
	if (clonedRequest.method !== "GET" && clonedRequest.body !== null && !("duplex" in clonedRequest)) clonedRequest.duplex = "half";
	return clonedRequest;
};

//#endregion
export { IsOptIntoMiddleware, assertValidHandlerResult, getResponseClerkState, injectRequestStateIntoResponse, isDataWithResponseInit, isRedirect, isResponse, parseCookies, patchRequest, wrapWithClerkState };
