// @clerk/react-router/dist/client/ReactRouterClerkProvider.js
import { assertPublishableKeyInSpaMode, assertValidClerkState, isSpaMode, warnForSsr } from "../utils/assert.js";
import { getPublicEnvVariables } from "../utils/env.js";
import { ClerkReactRouterOptionsProvider } from "./ReactRouterOptionsContext.js";
import { useAwaitableNavigate } from "./useAwaitableNavigate.js";
import { InternalClerkProvider } from "@clerk/react/internal";
import React from "react";

//#region src/client/ReactRouterClerkProvider.tsx
const SDK_METADATA = {
	name: "@clerk/react-router",
	version: "3.4.5"
};
/**
* React Router hydration errors should not stop Clerk navigation from working, as the components mount only after
* hydration is done (in the case of a hydration error, the components will simply mount after client-side hydration).
*/
const awaitableNavigateRef = { current: void 0 };
function ClerkProviderBase({ children, ...rest }) {
	const awaitableNavigate = useAwaitableNavigate();
	const isSpaMode$1 = isSpaMode();
	React.useEffect(() => {
		awaitableNavigateRef.current = awaitableNavigate;
	}, [awaitableNavigate]);
	const { clerkState, ...restProps } = rest;
	InternalClerkProvider.displayName = "ReactClerkProvider";
	if (typeof isSpaMode$1 !== "undefined" && !isSpaMode$1) assertValidClerkState(clerkState);
	const { __clerk_ssr_state, __publishableKey, __proxyUrl, __domain, __isSatellite, __clerk_debug, __signInUrl, __signUpUrl, __signInForceRedirectUrl, __signUpForceRedirectUrl, __signInFallbackRedirectUrl, __signUpFallbackRedirectUrl, __clerkJSUrl, __clerkJSVersion, __clerkUIUrl, __clerkUIVersion, __prefetchUI, __telemetryDisabled, __telemetryDebug, __unsafeDisableDevelopmentModeConsoleWarning, __keylessClaimUrl, __keylessApiKeysUrl } = clerkState?.__internal_clerk_state || {};
	React.useEffect(() => {
		if (typeof isSpaMode$1 !== "undefined" && !isSpaMode$1) warnForSsr(clerkState);
	}, []);
	React.useEffect(() => {
		window.__clerk_debug = __clerk_debug;
	}, []);
	const mergedProps = {
		publishableKey: __publishableKey,
		proxyUrl: __proxyUrl,
		domain: __domain,
		isSatellite: __isSatellite,
		signInUrl: __signInUrl,
		signUpUrl: __signUpUrl,
		signInForceRedirectUrl: __signInForceRedirectUrl,
		signUpForceRedirectUrl: __signUpForceRedirectUrl,
		signInFallbackRedirectUrl: __signInFallbackRedirectUrl,
		signUpFallbackRedirectUrl: __signUpFallbackRedirectUrl,
		__internal_clerkJSUrl: __clerkJSUrl,
		__internal_clerkJSVersion: __clerkJSVersion,
		__internal_clerkUIUrl: __clerkUIUrl,
		__internal_clerkUIVersion: __clerkUIVersion,
		prefetchUI: __prefetchUI,
		telemetry: {
			disabled: __telemetryDisabled,
			debug: __telemetryDebug
		},
		unsafe_disableDevelopmentModeConsoleWarning: __unsafeDisableDevelopmentModeConsoleWarning ?? getPublicEnvVariables(void 0).unsafeDisableDevelopmentModeConsoleWarning
	};
	const keylessProps = __keylessClaimUrl ? {
		__internal_keyless_claimKeylessApplicationUrl: __keylessClaimUrl,
		__internal_keyless_copyInstanceKeysUrl: __keylessApiKeysUrl
	} : {};
	return /* @__PURE__ */ React.createElement(ClerkReactRouterOptionsProvider, { options: mergedProps }, /* @__PURE__ */ React.createElement(InternalClerkProvider, {
		routerPush: (to) => awaitableNavigateRef.current?.(to),
		routerReplace: (to) => awaitableNavigateRef.current?.(to, { replace: true }),
		initialState: __clerk_ssr_state,
		sdkMetadata: SDK_METADATA,
		...mergedProps,
		...keylessProps,
		...restProps
	}, children));
}
const ClerkProvider = ({ children, loaderData, ...opts }) => {
	let clerkState;
	const isSpaMode$2 = isSpaMode();
	if (!isSpaMode$2 && loaderData?.clerkState) clerkState = loaderData.clerkState;
	if (typeof isSpaMode$2 !== "undefined" && isSpaMode$2) assertPublishableKeyInSpaMode(opts.publishableKey);
	return /* @__PURE__ */ React.createElement(ClerkProviderBase, {
		...opts,
		clerkState
	}, children);
};

//#endregion
export { ClerkProvider };
