// @clerk/react-router/dist/server/keyless/index.js
import { clerkClient } from "../clerkClient.js";
import { createFileStorage } from "./fileStorage.js";
import { createKeylessService } from "@clerk/shared/keyless";

//#region src/server/keyless/index.ts
let keylessServiceInstance = null;
function keyless(args, options) {
	if (!keylessServiceInstance) keylessServiceInstance = createKeylessService({
		storage: createFileStorage(),
		api: {
			async createAccountlessApplication(requestHeaders, source) {
				try {
					return await clerkClient(args, options).__experimental_accountlessApplications.createAccountlessApplication({
						requestHeaders,
						source
					});
				} catch {
					return null;
				}
			},
			async completeOnboarding(requestHeaders, source) {
				try {
					return await clerkClient(args, options).__experimental_accountlessApplications.completeAccountlessApplicationOnboarding({
						requestHeaders,
						source
					});
				} catch {
					return null;
				}
			}
		},
		framework: "react-router"
	});
	return keylessServiceInstance;
}

//#endregion
export { keyless };
